### PR TITLE
trial comparisons page

### PR DESCRIPTION
**Pull request by emilia-friedberg**
## WHAT
Build the Trial Comparisons page in the Evidence CMS.

## WHY
This translates over the feature Brendan built in ERB.

## HOW
Updating the comparisons controller to enable JSON versions of the create/show actions, then update the frontend to actually call the creation and redirect to a new page that shows the data. This works pretty similarly to the `EvaluationSection` from the `IndividualTrial` page, so I abstracted out some shared logic to build the `ComparisonsSection`.

Bonus: updated `DataTable` to better handle aria roles for its checkboxes, which updated some additional snapshots.

Note: I updated the `comparisons` factory to build the trials, since it's not really testable without it, but if there's a better method to use here I'm very open to changing that.

### Screenshots
Screenshot 2024-11-13 at 3 51 57 PM

### Notion Card Links
https://www.notion.so/quill/Trial-Comparisons-Page-13cd42e6f941808289f7d72c64818120?pvs=4

### What have you done to QA this feature?
Tested locally and having Jamie test on staging.

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? | YES
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | YES

### CODE REVIEW

File: services/QuillLMS/engines/evidence/spec/factories/evidence/research/gen_ai/comparisons.rb
Comment (brendanshean) [inline services/QuillLMS/engines/evidence/spec/factories/evidence/research/gen_ai/comparisons.rb:1]: I think some of our after create callbacks on factories are a little too opinionated, but I think this one is correct.  Comparisons don't really have any value without trials.

**Review by brendanshean — APPROVED** [review]
This looks great!  Really nice to see all of these pieces pulled together.

File: services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/individualDataset/trialsSection.tsx [hunk 1]
```diff
@@ -4,11 +4,22 @@ import moment from 'moment';
...
+import { createComparison } from '../../../../utils/evidence/genAIAPIs';
 
-const TrialsSection = ({ trials, trialsPath, }: { trials: TrialInterface[], trialsPath: string }) => {
+const TrialsSection = ({ trials, datasetPath, }: { trials: TrialInterface[], datasetPath: string }) => {
   const [trialIdsToCompare, setTrialIdsToCompare] = React.useState<number[]>([])
 
-  function compareTrials() { window.alert('Comparison coming soon!') }
+  function compareTrials() {
+    createComparison(trials[0].dataset_id, trialIdsToCompare, comparisonCreationSuccessFunction, comparisonCreationErrorFunction)
+  }
```
Comment (eadams17) [inline services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/individualDataset/trialsSection.tsx:14]: given the number of arguments, maybe we should use an object here?

File: services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/individualTrial/individualTrial.tsx [hunk 2]
```diff
@@ -7,9 +7,9 @@ import TrialNotesModal from './trialNotesModal'
 import ViewLLMPromptModal from './viewLLMPromptModal'
 import ViewGEvalModal from './viewGEvalModal'
 import Histogram from './histogram'
-import ConfusionMatrix from './confusionMatrix'
 import EvaluationSection from './evaluationSection'
 
+import ConfusionMatrix from '../shared/confusionMatrix'
```
Comment (eadams17) [inline services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/individualTrial/individualTrial.tsx:12]: I'm always going to love this name

File: services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/shared/sharedFunctions.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+export function llmExampleMatchColor(llmExample: LLMExampleInterface, taskType: string) {
+  if (matched(llmExample, taskType)) { return LIGHT_GREEN}
+
+  if (llmExample['optimal?']) { return LIGHT_RED }
+  if (llmExample['suboptimal?']) { return DARK_RED }
+}
+
+export function matched(llmExample: LLMExampleInterface, taskType: string) {
+  if (taskType === CLASSIFICATION && llmExample.rag_label == llmExample.test_example.rag_label) {
```
Comment (eadams17) [inline services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/shared/sharedFunctions.ts]: I think we want a triple equal comparison on the right here too?

File: services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/trialComparison/comparisonsSection.tsx [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        <a href={`${trialPath}/${id}`}>Trial: #{number}</a>
+        <span>Model: {llm_version}</span>
+        <span>Guidelines: {optimal_guidelines_count} Optimal, {suboptimal_guidelines_count} Sub-Optimal</span>
+        <span>Examples: {optimal_examples_count} Optimal, {suboptimal_examples_count} Sub-Optimal</span>
+        <span>Prompt Template: {llm_prompt_template.name}</span>
+        <span>Temperature: {temperature}</span>
+        <ConfusionMatrix confusionMatrix={results.confusion_matrix} />
+      </div>
+    )
+  }
```
Comment (eadams17) [inline services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/trialComparison/comparisonsSection.tsx:96]: love the structuring of logic here

File: services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/trialComparison/comparisonsSection.tsx [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+    const relevantLLMExamples = llmExamples.filter(llmExample => llmExample.test_example_id === testExample.id)
+
+    relevantLLMExamples.forEach(llmExample => {
+      const { llm_feedback, trial_id } = llmExample
+      const trial = trials.find(t => t.id === trial_id)
+      const gEval = trial.results.g_evals && Object.keys(trial.results.g_evals).map((gEvalId) => parseInt(trial.results.g_evals[gEvalId][index]) || '')
+      rowAttributes[`trial${trial.id}`] = (
+        <div className="llm-example-cell">
+          <span className={`match-square ${llmExampleMatchColor(llmExample, task_type)}`}>{llmExample['optimal?'] ? 'O' : 'S'}</span>
```
Comment (eadams17) [inline services/QuillLMS/client/app/bundles/Staff/components/evidence/llmPromptDatasets/trialComparison/comparisonsSection.tsx]: just curious what 0 v.s. S means here?

**Review by eadams17 — APPROVED** [review]
Looking great! Just a few small comments

**Review by anathomical — APPROVED** [review]

**Pull request merged**